### DISCUSSION TITLE
ConnectionBase.cxx:754 | HTTP/1.1 400 Bad Request  Sec-WebSocket-Version: 13

### DISCUSSION THREAD

**Discussion by coolbreeze1093** [resiprocate/resiprocate#317] [Q&A]
Is this error missing some configuration? Help me! 
<img width="1255" alt="image" src="https://github.com/resiprocate/resiprocate/assets/101555502/348aaf37-538c-4184-89b3-49bba7f5c35d">
<img width="1260" alt="image" src="https://github.com/resiprocate/resiprocate/assets/101555502/9783d75f-5e6b-4fc8-83bb-7798fed164eb">

**Comment by sgodin** [comment]
Look at the line starting with ERR, you are getting an error loading a certificate file:  domain_cert_server.pem.  I assume the server certificate is requesting a client certificate.

Scott

**Reply by coolbreeze1093** [reply]
The certificate is loaded successfully, but the header of the websocket is wrong, which may be the reason for accessing freeswitch

**Reply by coolbreeze1093** [reply]
https://github.com/resiprocate/resiprocate/issues/173

**Reply by coolbreeze1093** [reply]
sipstack is really too complicated, I don't know how to fix the problem of websocket handshake failure, then delete sipstack and introduce websocketpp. This is a very bad practice, because there is a timer function in sipstack, I have to reimplement it, it is really too complicated

**Reply by coolbreeze1093** [reply]
![image](https://github.com/resiprocate/resiprocate/assets/101555502/8d6387b4-ec18-4f86-be56-46e2d935da6f)

**Reply by coolbreeze1093** [reply]
This is the packet capture message of ws requesting freeswitch. From this analysis, the problem should be that websocket did not execute the handshake request.

![image](https://github.com/resiprocate/resiprocate/assets/101555502/3e015c51-aae3-4383-a6f2-c7dd0e927c3f)
![image](https://github.com/resiprocate/resiprocate/assets/101555502/96dea3f8-2f17-4ddb-b25f-915f74e20778)

1. The initial handshake of Websocket needs to be completed with the help of HTTP request, which is also based on TCP, that is, the browser and server establish a TCP connection and shake hands three times.
2. After the TCP connection is successful, the browser transmits information such as the version number supported by WebSocket to the server through the HTTP protocol.
3. After receiving the handshake request from the client, the server also uses the HTTP protocol to feed back data.
4. After receiving the successful connection message, transmit and communicate through the TCP channel.

If this is a bug, hope it gets fixed soon！！！

**Comment by sgodin** [comment]
Thank you for this clear description.  I see code in Connection.cxx that seems to be where the client side WS handshake process should be occurring.  See:  Connection::performWrite..
   else if(mSendingTransmissionFormat == WebSocketHandshake)
   {
      mSendingTransmissionFormat = WebSocketData;
   }

I did not personally work on any of this code in the past, and I don't have any immediate need for client side websocket connections.  There is no one company behind resip with resources to fix issues, like this, that are identified by the community.  The only way this is going to get fixed is if someone in the community spends time on it.  We would gladly accept any pull requests to fix this.
